common/reactive/src/main/java/io/helidon/common/reactive/Mono.java

```diff
@@ -0,0 +1,202 @@
+import io.helidon.common.reactive.Flow.Subscriber;
+
+/**
+ * Single item publisher.
+ * @param <T> published type
+ */
+public interface Mono<T> extends Publisher<T> {
+
+    /**
+     * Retrieve the value of this {@link Mono} instance in a blocking manner.
+     *
+     * @return value
+     * @throws IllegalStateException if the Mono wraps an error or if
+     * interrupted
+     */
+    default T block() {
+        MonoBlockingSubscriber<T> subscriber = new MonoBlockingSubscriber<>();
+        this.subscribe(subscriber);
+        return subscriber.blockingGet();
+    }
+
+    /**
+     * Retrieve the value of this {@link Mono} instance in a blocking manner.
+     *
+     * @param timeout timeout value
+     * @return value
+     * @throws IllegalStateException if the Mono wraps an error, or the
+     * timeout is reached or if interrupted
+     */
+    default T block(Duration timeout) {
+        MonoBlockingSubscriber<T> subscriber = new MonoBlockingSubscriber<>();
+        this.subscribe(subscriber);
+        return subscriber.blockingGet(timeout.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Map this {@link Mono} instance to a new {@link Mono} of another type
+     * using the given {@link MonoMapper}.
+     * @param <U> mapped item type
+     * @param mapper mapper
+     * @return Mono
+     */
+    default <U> Mono<U> map(MonoMapper<T, U> mapper) {
+        this.subscribe(mapper);
+        return mapper;
+    }
+
+    /**
+     * Map this {@link Mono} instance to a new {@link Mono} of another type
+     * using the given java {@link Function}.
+     * @param <U> mapped item type
+     * @param function mapper function
+     * @return Mono
+     */
+    default <U> Mono<U> map(Function<T, U> function) {
+        MonoMapperFunctional<T, U> mapper = new MonoMapperFunctional<>(function);
+        this.subscribe(mapper);
+        return mapper;
+    }
+
+    /**
+     * Map this {@link Mono} instance to a publisher using the given
+     * {@link MonoMultiMapper}.
+     *
+     * @param <U> mapped items type
+     * @param mapper mapper
+     * @return Publisher
+     */
+    default <U> Publisher<U> mapMany(MonoMultiMapper<T, U> mapper) {
+        this.subscribe(mapper);
+        return mapper;
+    }
+
+    /**
+     * Map this {@link Mono} instance to a publisher using the given
+     * java {@link Function}.
+     *
+     * @param <U> mapped items type
+     * @param function mapper function
+     * @return Publisher
+     */
+    default <U> Publisher<U> mapMany(
+            Function<T, Publisher<U>> function) {
+
+        MonoMultiMapperFunctional<T, U> mapper =
+                new MonoMultiMapperFunctional<>(function);
+        this.subscribe(mapper);
+        return mapper;
+    }
+
+    /**
+     * Exposes this {@link Mono} instance as a {@link CompletableFuture}.
+     * @return CompletableFuture
+     */
+    default CompletableFuture<T> toFuture() {
+        try {
```

Comment:
The reason you suggested using `Optional<T>` was because the current implementation completes the future with `null` when the `Single`/`Mono` completes without value. Instead we can complete the future exceptionally with an exception.

`CompletableFuture<Optional<T>>` feels too much impractical for most use-cases.